### ISSUE TITLE
[Bug] Cannot install SCX

### ISSUE THREAD

**Issue by VeilSilence** [chaotic-cx/nyx#790]
### What happens?

Greetings. 

I'm using my custom kernel.
I put `chaotic.scx.enable = true;` in my configuration. Error log on spoiler.

<details><summary>Build error</summary>
<p>

```nix

error:
       … while calling the 'head' builtin
         at /nix/store/dk2rpyb6ndvfbf19bkb2plcz5y3k8i5v-source/lib/attrsets.nix:1575:11:
         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'
         at /nix/store/dk2rpyb6ndvfbf19bkb2plcz5y3k8i5v-source/lib/modules.nix:809:9:
          808|     in warnDeprecation opt //
          809|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          810|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/dk2rpyb6ndvfbf19bkb2plcz5y3k8i5v-source/nixos/modules/system/activation/top-level.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error:
       Failed assertions:
       - SCX needs a kernel compiled with CONFIG_BPF
       - SCX needs a kernel compiled with CONFIG_BPF_EVENTS
       - SCX needs a kernel compiled with CONFIG_BPF_JIT
       - SCX needs a kernel compiled with CONFIG_BPF_SYSCALL
       - SCX needs a kernel compiled with CONFIG_DEBUG_INFO_BTF
       - SCX needs a kernel compiled with CONFIG_FTRACE
       - SCX needs a kernel compiled with CONFIG_SCHED_CLASS_EXT
┏━ 1 Errors:
 ⋮
┃        … while evaluating the option `system.build.toplevel':
┃
┃        … while evaluating definitions from `/nix/store/dk2rpyb6ndvfbf19bkb2pl…
┃
┃        (stack trace truncated; use '--show-trace' to show the full, detailed …
┃
┃        error:
┃        Failed assertions:
┃        - SCX needs a kernel compiled with CONFIG_BPF
┃        - SCX needs a kernel compiled with CONFIG_BPF_EVENTS
┃        - SCX needs a kernel compiled with CONFIG_BPF_JIT
┃        - SCX needs a kernel compiled with CONFIG_BPF_SYSCALL
┃        - SCX needs a kernel compiled with CONFIG_DEBUG_INFO_BTF
┃        - SCX needs a kernel compiled with CONFIG_FTRACE
┃        - SCX needs a kernel compiled with CONFIG_SCHED_CLASS_EXT
┣━━━
┗━ 󰒠  Exited with 1 errors reported by nix at 20:44:00 after 6s
Error:
   0: Command exited with status Exited(1)
```
</p>
</details> 

<details><summary>My Kernel config</summary>
<p>

```shell

  zcat /proc/config.gz | grep -i "CONFIG_BPF"
CONFIG_BPF=y
CONFIG_BPF_SYSCALL=y
CONFIG_BPF_JIT=y
CONFIG_BPF_JIT_ALWAYS_ON=y
CONFIG_BPF_JIT_DEFAULT_ON=y
CONFIG_BPF_UNPRIV_DEFAULT_OFF=y
# CONFIG_BPF_PRELOAD is not set
CONFIG_BPF_LSM=y
CONFIG_BPF_STREAM_PARSER=y
# CONFIG_BPF_LIRC_MODE2 is not set
CONFIG_BPF_EVENTS=y

❯  zcat /proc/config.gz | grep -i "CONFIG_BPF_EVENTS"
CONFIG_BPF_EVENTS=y

❯  zcat /proc/config.gz | grep -i "CONFIG_BPF_JIT"
CONFIG_BPF_JIT=y
CONFIG_BPF_JIT_ALWAYS_ON=y
CONFIG_BPF_JIT_DEFAULT_ON=y


❯  zcat /proc/config.gz | grep -i "CONFIG_BPF_SYSCALL"
CONFIG_BPF_SYSCALL=y


❯  zcat /proc/config.gz | grep -i "CONFIG_DEBUG_INFO_BTF"
CONFIG_DEBUG_INFO_BTF=y
CONFIG_DEBUG_INFO_BTF_MODULES=y


❯  zcat /proc/config.gz | grep -i "CONFIG_FTRACE"
CONFIG_FTRACE=y
CONFIG_FTRACE_SYSCALLS=y
CONFIG_FTRACE_MCOUNT_RECORD=y
CONFIG_FTRACE_MCOUNT_USE_OBJTOOL=y
# CONFIG_FTRACE_RECORD_RECURSION is not set
# CONFIG_FTRACE_STARTUP_TEST is not set
# CONFIG_FTRACE_SORT_STARTUP_TEST is not set


❯  zcat /proc/config.gz | grep -i "CONFIG_SCHED_CLASS_EXT"
CONFIG_SCHED_CLASS_EXT=y
```

</p>
</details> 

### What is expected to happen?

Package succesfully installed.

### More information

I tried to install CCX with `nix run github:chaotic-cx/nyx/nyxpkgs-unstable#scx` out of curiosity.
I can launch them with sudo without errors.
Any ideas how to fix this issue?

**Comment by PedroHLC** [comment]
You need to set a kernel (`boot.kernelPackages`) that has the sched-ext patches applied and enabled. This repo provides `linuxPackages_cachyos` that fits.

**Comment by PedroHLC** [comment]
Your kernel needs to expose the configs in the nix-way. In `linuxManualConfig` that's done in the `config` attr: https://github.com/chaotic-cx/nyx/blob/5cd32d60b3f8ce3b0552103bb8de0635c3c0d87d/pkgs/linux-cachyos/kernel.nix#L23

**Comment by VeilSilence** [comment]
Hm.. I see.
 I'm just overriding kernel and applying patches like that for example:
<details><summary>Details</summary>
<p>

```nix
{lib, ...}: let
  inherit (lib.kernel) yes no freeform;
  inherit (lib.attrsets) mapAttrs;
  inherit (lib.modules) mkForce;
in {
  boot.kernelPatches = [
    {
      # EXT sched patch
      name = "ext";
      patch = ./patches/bore-cachy-ext/0001-cachy-base-sched.patch;
    }

    {
      # bore sched
      name = "bore";
      patch = ./patches/bore-cachy-ext/0001-Bore-cachy-ext.patch;
      extraStructuredConfig = mapAttrs (_: mkForce) {
        # imported from here: https://github.com/CachyOS/linux-cachyos/blob/master/linux-cachyos-bore/PKGBUILD
        SCHED_CLASS_EXT = yes;
        MIN_BASE_SLICE_NS = freeform "1000000";
        SCHED_BORE = yes;
        CACHY = yes;
      };
    }
  ];
}


```

</p>
</details>

**Comment by PedroHLC** [comment]
Then make sure the mentioned configs are set at `extraStructuredConfig`

**Comment by PedroHLC** [comment]
@VeilSilence is this one solved? Can I close it?

**Comment by VeilSilence** [comment]
I tried to find the way to make it work, but alas, without success.

<details><summary>Default.nix</summary>
<p>

```nix

{
  config,
  pkgs,
  ...
}: let
  inherit (config.networking) hostname;
  inherit (pkgs.callPackage ./xanmod-scx.nix {inherit hostname;}) xanmod-scx;
in {
  imports = [
    ./config
  ];
  config = {
  modules.system.boot.kernel = pkgs.linuxPackagesFor xanmod-scx;
  };
}

```

</p>
</details> 

<details><summary>xanmod-scx.nix</summary>
<p>

```nix

{
  lib,
  fetchFromGitHub,
  linuxKernel,
  callPackage,
  pkgs,
  config,
  hostname ? "",
  ...
} @ inputs: let
  inherit (lib.kernel) yes no freeform;
  inherit (lib.attrsets) mapAttrs;
  inherit (lib.modules) mkForce;

  # calling llvm helper
  stdenvLLVM = callPackage ./lib/llvm-stdenv.nix {};

  version = "6.9.8";
  suffix = "xanmod1";
  modDirVersion = "${version}-${suffix}";

  xanmod_custom = linuxKernel.kernels.linux_xanmod_latest.override {
    inherit version suffix modDirVersion;
    stdenv = stdenvLLVM;

    # https://github.com/xanmod/linux
    src = fetchFromGitHub {
      owner = "xanmod";
      repo = "linux";
      rev = "refs/tags/${version}-xanmod1";
      hash = "sha256-rSwvVt+DW7rc//aY6Ho8mllUy/3MKuAs/71UG+p07Iw=";
    };

    ignoreConfigErrors = true;

in {
  inherit xanmod-scx;
}

```

</p>
</details> 

In `config`directory, `default.nix` import all my changed to kernel setting including:

<details><summary>./config/xanmod-sched-ext.nix</summary>
<p>

```nix
{lib, ...}: let
  inherit (lib.kernel) yes no freeform;
  inherit (lib.attrsets) mapAttrs;
  inherit (lib.modules) mkForce;
in {
  boot.kernelPatches = [
    {
      # rust sched ext
      name = "sched-ext";
      patch = ./patches/Xanmod-sched-ext/0001-xanmod-sched-ext.patch;
      extraStructuredConfig = {
        # imported from here: https://github.com/CachyOS/linux-cachyos/blob/master/linux-cachyos-bore/PKGBUILD
        SCHED_CLASS_EXT = yes;
        FTRACE = yes;
        DEBUG_INFO_BTF = yes;
        BPF_SYSCALL = yes;
        BPF_JIT = yes;
        BPF_EVENTS = yes;
        BPF = yes;
      };
    }
  ];
}

```

</p>
</details> 

I'm very noobie in nix, so I'm missing something very obvious here.

**Comment by PedroHLC** [comment]
Since I'm traveling I cannot debug it myself, you might open `nix repl <your-flake>` and from there find the `nixosConfigurations.<your-hostname>.config.boot.kernelPackages.kernel.passthru.config` attrset, and inspect it to see what could be wrong...

**Comment by VeilSilence** [comment]
Thanks neat way to check it, thank you so much.

<details><summary>nix-repl</summary>
<p>

```nix
nixosConfigurations.Nix.config.boot.kernelPackages.kernel.passthru.commonStructuredConfig.SCHED_CLASS_EXT
error:
       … while evaluating the attribute 'Nix.config.boot.kernelPackages.kernel.passthru.commonStructuredConfig.SCHED_CLASS_EXT'
         at /nix/store/dk2rpyb6ndvfbf19bkb2plcz5y3k8i5v-source/pkgs/os-specific/linux/kernel/generic.nix:232:13:
          231|     features = kernelFeatures;
          232|     inherit commonStructuredConfig structuredExtraConfig extraMakeFlags isZen isHardened isLibre;
             |             ^
          233|     isXen = lib.warn "The isXen attribute is deprecated. All Nixpkgs kernels that support it now have Xen enabled." true;

       error: attribute 'SCHED_CLASS_EXT' missing
       at «string»:1:1:
            1| nixosConfigurations.Nix.config.boot.kernelPackages.kernel.passthru.commonStructuredConfig.SCHED_CLASS_EXT
             | ^
             
```

</p>
</details> 

```nix
zcat /proc/config.gz | grep -i "SCHED_CLASS_EXT"
CONFIG_SCHED_CLASS_EXT=y
```

```nix

boot.kernelPatches = [
    {
      # rust sched ext
      name = "sched-ext";
      patch = ./patches/Xanmod-sched-ext/0001-xanmod-sched-ext.patch;
      extraStructuredConfig = {
        # imported from here: https://github.com/CachyOS/linux-cachyos/blob/master/linux-cachyos-bore/PKGBUILD
        "SCHED_CLASS_EXT" = yes;
        FTRACE = yes;
        DEBUG_INFO_BTF = yes;
        BPF_SYSCALL = yes;
        BPF_JIT = yes;
        BPF_EVENTS = yes;
        BPF = yes;
      };
    }
  ];
  
```

So, what do I did wrong then, I don't get it. Not properly applied scx-ext.patch or what.

**Comment by PedroHLC** [comment]
Isn't it `CONFIG_SCHED_CLASS_EXT` instead of `SCHED_CLASS_EXT`?

**Comment by VeilSilence** [comment]
> Isn't it `CONFIG_SCHED_CLASS_EXT` instead of `SCHED_CLASS_EXT`?

I tried that as well.
I checked with `nix repl` and it looks like `extraStructuredConfig` is passing kernel flags directly into `configfile.drv`.
The `configfile.drv` changing, when I'm adding a new kernel options, but `config`, `commonStructuredConfig`, and `structuredExtraConfig` remain the same regardless of how many kernel options I enable.

We can probably close this issue.
